jetcd-core/src/main/java/io/etcd/jetcd/ClientBuilder.java

```diff
@@ -142,6 +143,36 @@ public ClientBuilder password(ByteSequence password) {
     return this;
   }
 
+  public ByteSequence namespace() {
+    return namespace;
+  }
+
+  /**
+   * config the namespace of keys used in {@code KV}, {@code Txn}, {@code Lock} and {@code Watch}.
+   * "/" will be treated as no namespace.
+   *
+   * @param namespace the namespace of each key used
+   * @return this builder
+   * @throws NullPointerException if namespace is <code>null</code>
+   */
+  public ClientBuilder namespace(ByteSequence namespace) {
+    checkNotNull(namespace);
+    if (namespace.getByteString().isEmpty()) {
+      this.namespace = ByteSequence.EMPTY;
+    } else if (namespace.size() == 1
+        && namespace.getByteString().equals(ByteSequence.NAMESPACE_DELIMITER.getByteString())) {
+      // treat "/" as no namespace
+      this.namespace = ByteSequence.EMPTY;
+    } else if (namespace.getByteString().byteAt(namespace.getByteString().size() - 1) == '/') {
+      this.namespace = namespace;
+    } else {
+      // make sure the namespace ends with '/'
+      this.namespace = ByteSequence.from(namespace.getByteString().concat(
+          ByteSequence.NAMESPACE_DELIMITER.getByteString()));
```

Comment:
I think you might miss the previous condition, if the namespace ends with '/', do nothing.
`    } else if (namespace.getByteString().byteAt(namespace.getByteString().size() - 1) == '/') {
      this.namespace = namespace;
    }`

> Also I am unsure if we should enforce the ending the namespace with '/' since etcd go client doesn't do it if I recall correctly.
Yes, I've noticed that.In etcd, suppose a given key a, etcd will treat a+1 as the excluded end range.
I did it on purpose since I though it might be error-prone for users if we don't enforce the ending character. Suppose there are two namespace 'a' and 'aa':
We will get all the keys inside namespace 'aa' once we try to list all the keys inside namespace 'a' if we don't ending the ending '/'.
What do you think of it ?

Reply:
That's good edge case. However, I don't think adding '/' will solve the issue. Suppose that there are two namespaces `a/` and `a//`. what if I do a range GET for all the keys with the name space `a/` and with the range end be`a/ + 1` , I would also get all the keys from `a//` right?

Reply:
I think for now we should stick with etcd implementation and file an issue on etcd to discuss the edge case that you were thinking about. Once we reach a  conclusion, then we can have another pr to update jetcd.

Reply:
Got it, I'll change this part in the next version.